File: 57-insert-interval/insert-interval.py
class Solution:
    def insert(self, intervals: List[List[int]], newInterval: List[int]) -> List[List[int]]:
        result = []
        for interval in intervals:
            if interval[1] < newInterval[0]:
                result.append(interval)
            elif interval[0] > newInterval[1]:
                result.append(newInterval)
                newInterval = interval
            elif interval[1] >= newInterval[0] or interval[0] <= newInterval[1]:
                newInterval[0] = min(interval[0], newInterval[0])
                newInterval[1] = max(interval[1], newInterval[1])
        else:
            result.append(newInterval)
        return result
        
        """
        intervals.append(newInterval)
        intervals.sort(key=lambda x: x[0])
        ans = [intervals[0]]
        for i in range(1, len(intervals)):
            a = ans[-1]
            b = intervals[i]
            if a[1] >= b[0]:
                ans[-1][1] = max(a[1], b[1])
            else:
                ans.append(b)
        return ans
        """

File: 57-insert-interval/insert-interval.java
class Solution {
    public int[][] insert(int[][] intervals, int[] newInterval) {
        int start = newInterval[0], end = newInterval[1];
        int i = 0, n = intervals.length, f = -1;
        while (i < n && start > intervals[i][1]) {
            f = i;
            i++;
        }
        while (i < n && intervals[i][0] <= end) {
            start = Math.min(start, intervals[i][0]);
            end = Math.max(end, intervals[i][1]);
            i++;
        }
        int[][] result = new int[f + 2 + n - i][2];
        int ci = 0;
        while (ci <= f) {
            result[ci] = intervals[ci];
            ci++;
        }
        result[ci++] = new int[] { start, end };
        while (i < n) {
            result[ci] = intervals[i];
            ci++;
            i++;
        }
        return result;
    }
}
